drivers/syslog/syslog_device.c

```diff
@@ -451,38 +451,58 @@ static ssize_t syslog_dev_write(FAR struct syslog_channel_s *channel,
 
       if (*endptr == '\r' || *endptr == '\n')
         {
+          /* Write everything up to the position of the special
+           * character.
+           *
+           * - buffer points to next byte to output.
+           * - endptr points to the special character.
+           */
+
+          writelen = (size_t)((uintptr_t)endptr - (uintptr_t)buffer);
+          if (writelen > 0)
+            {
+              nwritten = file_write(&syslog_dev->sl_file,
+                                    buffer, writelen);
+              if (nwritten < 0)
+                {
+                  ret = (int)nwritten;
+                  goto errout_with_sem;
+                }
+            }
+
           /* Check for pre-formatted CR-LF sequence */
 
           if (remaining > 1 &&
               ((endptr[0] == '\r' && endptr[1] == '\n') ||
                (endptr[0] == '\n' && endptr[1] == '\r')))
             {
-              /* Just skip over pre-formatted CR-LF or LF-CR sequence */
+              nwritten = file_write(&syslog_dev->sl_file,
+                                    g_syscrlf, 2);
 
-              endptr++;
-              remaining--;
-            }
-          else
-            {
-              /* Write everything up to the position of the special
-               * character.
-               *
-               * - buffer points to next byte to output.
-               * - endptr points to the special character.
+              /* Synchronize the file when each CR-LF is encountered
+               * (i.e., implements line buffering always).
                */
 
-              writelen = (size_t)((uintptr_t)endptr - (uintptr_t)buffer);
-              if (writelen > 0)
+#ifndef CONFIG_DISABLE_MOUNTPOINT
+              if (nwritten > 0)
                 {
-                  nwritten = file_write(&syslog_dev->sl_file,
-                                        buffer, writelen);
-                  if (nwritten < 0)
-                    {
-                      ret = (int)nwritten;
-                      goto errout_with_sem;
-                    }
+                  syslog_dev_flush(channel);
```

Comment:
we can merge the most code in if and else block:
```
nwritten = file_write(&syslog_dev->sl_file, g_syscrlf, 2);
if (nwritten < 0)
  {
    ret = (int)nwritten;
    goto errout_with_sem;
  }

syslog_dev_flush(channel);

if (remaining > 1 &&
    ((endptr[0] == '\r' && endptr[1] == '\n') ||
    (endptr[0] == '\n' && endptr[1] == '\r')))
  {
    endptr++;
    remaining--;
  }
```

Reply:
I do not see how this can be done. The `nwritten = file_write(&syslog_dev->sl_file, g_syscrlf, 2);` should be called either in case of `if (remaining > 1 && ((endptr[0] == '\r' && endptr[1] == '\n') || (endptr[0] == '\n' && endptr[1] == '\r')))` or in case of `if (*endptr == '\n')`, but not for `if (*endptr == '\r')`. Also `endptr++;` and `remaining--;` should be done only if we are having pre-formatted CR-LF sequence.
If you can come with a suggestion, then I will analyze and integrate it into this PR

Reply:
I can try to do:
```
          /* Check for pre-formatted CR-LF sequence */

          if (remaining > 1 &&
              ((endptr[0] == '\r' && endptr[1] == '\n') ||
               (endptr[0] == '\n' && endptr[1] == '\r')))
            {
              writelen = sizeof(g_syscrlf);

              /* Skip over pre-formatted CR-LF or LF-CR sequence */

              endptr++;
              remaining--;
            }
          else
            {
              /* Ignore the carriage return, but for the linefeed, output
               * both a carriage return and a linefeed.
               */

              if (*endptr == '\n')
                {
                  writelen = sizeof(g_syscrlf);
                }
              else
                {
                  writelen = 0;
                }
            }

            if (writelen > 0)
              {
                nwritten = file_write(&syslog_dev->sl_file,
                                      g_syscrlf, writelen);

                /* Synchronize the file when each CR-LF is encountered
                 * (i.e., implements line buffering always).
                 */

                 if (nwritten > 0)
                  {
                    syslog_dev_flush(channel);
                  }

                if (nwritten < 0)
                  {
                    ret = (int)nwritten;
                    goto errout_with_sem;
                  }
              }
```
@xiaoxiang781216 what do you think?

Reply:
Yes, it's much better than the previous version